src/hazelcore/semantics/UHPat.re

```diff
@@ -153,6 +158,51 @@ and make_inconsistent_operand =
     (Parenthesized(set_p), u_gen);
   };
 
+let patterns_of_type =
+    (u_gen: MetaVarGen.t, ty: HTyp.t): (list(t), MetaVarGen.t) =>
+  OpSeq.(
+    switch (ty) {
+    | HTyp.Hole
+    | Unit
+    | Arrow(_, _) => ([], u_gen)
+    | Num => ([numlit(0), wild()] |> List.map(wrap), u_gen)
+    | Bool => ([true, false] |> List.map(b => wrap(boollit(b))), u_gen)
+    | Prod(_, _) =>
+      let prod_element_count = ty |> HTyp.get_prod_elements |> List.length;
+      let patternSkel =
+        ListUtil.range(prod_element_count)
+        |> List.map(i => Skel.Placeholder(i))
+        |> make_tuple(NotInHole);
+      let opseq_of_holes = u_gen => {
```

Comment:
did you reach out to Qiran to ask about his tuple completion code? I imagine it has very similar logic in it, so you might be able to put it in a common helper function instead of duplicating it.

Reply:
I found his code (`Action._complete_tuple` I think`, but it think it looks a lot more complicated. I'll see if I can unify the two.
It seems construction of tuples with more than 2 elements is a little bit broken at the moment, but I'll ask about this in Slack.
Personally, I think interpreting nested `UHTyp.Prod`s as n-tuples could lead to undesired behaviour. For example, if I have a tuple `1, (2,3,4,5,6)`, I might want to just pattern match on the first element (`1, _`), but now I'm forced to have the same number of elements in my pattern (`1, _,_,_,_,_`). To fix this `UHTyp.t` has to be changed to contain information to distinguish a right element from being the rest of the tuple (for the n-tuple case) or a nested tuple as last element of the current tuple, but I think this would be out of scope for this PR.

Reply:
the nested product types are interpreted as n-tuples only if the right-hand side isn't parenthesized, so in your example if the type is (Num, (Num, Num, Num, Num, Num)), parenthesized like that, then it would be treated as a pair.

Reply:
But `HTyp` doesn't contain any information about the parentheses anymore, so doesn't that make it indistinguishable at that level? Also both `master` and `dev` currently expect a 6-tuple pattern.

Reply:
ah yeah that's true... maybe we should explicitly represent n-tuples in HTyp...

Reply:
yeah I think that makes sense. this is going to affect both your code and Qiran's tuple completion code. do you want to coordinate with him on doing this refactoring? (change Prod to take a list(t) argument + generate that list in the UHTyp.expand function + changing all the code that uses HTyp)

Reply:
Alright, I'll coordinate with him.

Reply:
@cdfa Keep me in the loop re: n-tuples. I recently refactored the typechecker (`Statics.re`) to support n-tuples, I can help you propagate the refactor there if you get stuck.